agentendpoint/config_task.go

```diff
@@ -0,0 +1,462 @@
+}
+
+type applyConfigTask struct {
+	*agentendpointpb.ApplyConfigTask
+}
+
+type assignment struct {
+	policies map[string]*policy
+}
+
+type policy struct {
+	resources map[string]*resource
+	hasError  bool
+}
+
+type resource struct {
+	enforcementNeeded bool
+
+	packageResource        *agentendpointpb.PackageResource
+	repositoryResource     *agentendpointpb.RepositoryResource
+	execResource           *agentendpointpb.ExecResource
+	fileResource           *agentendpointpb.FileResource
+	extractArchiveResource *agentendpointpb.ExtractArchiveResource
+	serviceResource        *agentendpointpb.ServiceResource
+}
+
+func (r *resource) unmarshal(ctx context.Context, res *agentendpointpb.ApplyConfigTask_Config_Resource) error {
+	// TODO: implement
+	return nil
+}
+
+func (r *resource) checkState(ctx context.Context) error {
+	// TODO: implement
+	r.enforcementNeeded = false
+	return nil
+}
+
+func (r *resource) runEnforcement(ctx context.Context) error {
+	// TODO: implement
+	return nil
+}
+
+func (c *configTask) reportCompletedState(ctx context.Context, errMsg string, state agentendpointpb.ApplyConfigTaskOutput_State) error {
+	req := &agentendpointpb.ReportTaskCompleteRequest{
+		TaskId:       c.TaskID,
+		TaskType:     agentendpointpb.TaskType_APPLY_CONFIG_TASK,
+		ErrorMessage: errMsg,
+		Output: &agentendpointpb.ReportTaskCompleteRequest_ApplyConfigTaskOutput{
+			ApplyConfigTaskOutput: &agentendpointpb.ApplyConfigTaskOutput{State: state, ConfigAssignmentResults: c.results},
+		},
+	}
+	if err := c.client.reportTaskComplete(ctx, req); err != nil {
+		return fmt.Errorf("error reporting completed state: %v", err)
+	}
+	return nil
+}
+
+func (c *configTask) handleErrorState(ctx context.Context, msg string, err error) error {
+	if err == errServerCancel {
+		clog.Infof(ctx, "Canceling config run: %v", errServerCancel)
+		return c.reportCompletedState(ctx, errServerCancel.Error(), agentendpointpb.ApplyConfigTaskOutput_CANCELLED)
+	}
+	msg = fmt.Sprintf("%s: %v", msg, err)
+	clog.Errorf(ctx, msg)
+	return c.reportCompletedState(ctx, msg, agentendpointpb.ApplyConfigTaskOutput_FAILED)
+}
+
+func (c *configTask) reportContinuingState(ctx context.Context, configState agentendpointpb.ApplyConfigTaskProgress_State) error {
+	req := &agentendpointpb.ReportTaskProgressRequest{
+		TaskId:   c.TaskID,
+		TaskType: agentendpointpb.TaskType_APPLY_CONFIG_TASK,
+		Progress: &agentendpointpb.ReportTaskProgressRequest_ApplyConfigTaskProgress{
+			ApplyConfigTaskProgress: &agentendpointpb.ApplyConfigTaskProgress{State: configState},
+		},
+	}
+	res, err := c.client.reportTaskProgress(ctx, req)
+	if err != nil {
+		return fmt.Errorf("error reporting task progress %s: %v", configState, err)
+	}
+	if res.GetTaskDirective() == agentendpointpb.TaskDirective_STOP {
+		return errServerCancel
+	}
+	return nil
+}
+
+func (c *configTask) validation(ctx context.Context) {
+	// First populate validation results and internal assignment state.
```

Comment:
I thought validation was ensuring the agent understands the resources and checks for conflicts. Mutating state here breaks the SRP. Should those two things (populating pre-enforcement, internal assignment state and validating the task) be separate functions?

Reply:
The assignment state here is required for validation steps (unmarshal, check conflicts), this 'state' is just this particular config tasks understanding of the given assignments, not any persistent state in the agent. For example, it needs to have a place to unmarshal the resources to so it can then check for conflicts.